File: Second year/Design and Programming Environments/TicketManagerCSharp/Server/ServerImpl.cs
using Persistence;
using Services;
using Services.Exceptions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using TicketManagerCSharp.domain;


namespace Server
{
    internal class ServerImpl : IServer
    {
        private UserRepository repositoryUser;
        private TicketRepository repositoryTicket;
        private EventRepository repositoryEvent;
        private readonly IDictionary<String, IObserver> loggedClients;

        public ServerImpl(UserRepository repositoryUser, TicketRepository repositoryTicket, EventRepository repositoryEvent)
        {
            this.repositoryUser = repositoryUser;
            this.repositoryTicket = repositoryTicket;
            this.repositoryEvent = repositoryEvent;
            loggedClients = new Dictionary<String, IObserver>();
        }

        public void addEvent(int id, string game_name, double game_price, int free_seats)
        {
            Event ev = new Event(id, game_name, game_price, free_seats);
            this.repositoryEvent.Save(ev);
            notify();
        }

        public void addTicket(int id_game, int reserved_seats, string client_name)
        {
            Event ev = repositoryEvent.Find(id_game);
            Ticket ticket = new Ticket(1, reserved_seats, reserved_seats * ev.GamePrice, client_name);
            if (ev.FreeSeats - ticket.ReservedSeats >= 0)
            {
                ev.FreeSeats = ev.FreeSeats - ticket.ReservedSeats;
                this.repositoryTicket.Save(ticket);
                updateEvent(ev.Id, ev.GameName, ev.GamePrice, ev.FreeSeats);
            }
        }

        public void deleteEvent(int id)
        {
            repositoryEvent.Delete(id);
            notify();
        }

        public void deleteTicket(int id)
        {
            Ticket ticket = repositoryTicket.Find(id);
            Event ev = repositoryEvent.Find(ticket.IdGame);
            ev.FreeSeats = ev.FreeSeats + ticket.ReservedSeats;
            repositoryTicket.Delete(id);
            updateEvent(ev.Id, ev.GameName, ev.GamePrice, ev.FreeSeats);
        }

        public IList<Event> getEvents()
        {
            return repositoryEvent.getALl();
        }

        public IList<Ticket> getTickets()
        {
            return repositoryTicket.getALl();
        }

        public void login(User user, IObserver client)
        {
            User currentUser = repositoryUser.FindOne(user.Id);
            if (currentUser.Id != null)
            {
                if (loggedClients.ContainsKey(currentUser.Id))
                    throw new BasketException("User already logged in");
                loggedClients[currentUser.Id] = client;
            }
            else
            {
                throw new BasketException("Authentication error");
            }
        }

        public void logout(User user)
        {
            IObserver localClient = loggedClients[user.Id];
            if (localClient == null)
                throw new BasketException("User " + user.Id + " is not logged in.");
            loggedClients.Remove(user.Id);
        }

        public void updateEvent(int id, string game_name, double game_price, int free_seats)
        {
            Event ev = new Event(id, game_name, game_price, free_seats);
            repositoryEvent.Update(id, ev);
            notify();
        }

        private void notify()
        {
            foreach (var obs in this.loggedClients.Values)
            {
                //Console.WriteLine("Impl Notified " + username);
                //this.loggedClients[username].update((List<Meci>)getMeciuri());
                Task.Run(() => obs.update((List<Event>)getEvents()));
            }
        }
    }
}

File: Second year/Design and Programming Environments/TicketManagerJava/Server/src/main/java/ServerImpl.java

import Exceptions.BasketException;
import Interfaces.IObserver;
import Interfaces.IServer;
import ModelDomain.Event;
import ModelDomain.Ticket;
import ModelDomain.User;
import Repository.EventRepository;
import Repository.TicketRepository;
import Repository.UserRepository;

import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class ServerImpl implements IServer {
    private TicketRepository ticketRepository;
    private EventRepository eventRepository;
    private UserRepository userRepository;

    private Map<String, IObserver> loggedPers;
    private List<IObserver> observers = new ArrayList<>();


    public ServerImpl(UserRepository userRepository, EventRepository eventRepository, TicketRepository ticketRepository) {
        this.userRepository = userRepository;
        this.eventRepository = eventRepository;
        this.ticketRepository = ticketRepository;

        loggedPers = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized void login(User person, IObserver client) throws BasketException {
        User user = userRepository.findOne(person.getId());
        if (user != null) {
            if (loggedPers.get(user.getId()) != null) {
                throw new BasketException("User already logged in!");
            }
            loggedPers.put(person.getId(), client);
        } else {
            throw new BasketException("Authentication error!");
        }
    }

    @Override
    public void logout(User person) throws BasketException {
        loggedPers.remove(person.getId());
    }

    @Override
    public Iterable<Ticket> getTickets() throws BasketException {
        return ticketRepository.findAll();
    }

    @Override
    public Iterable<Event> getEvents() throws BasketException {
        return eventRepository.findAll();
    }

    @Override
    public void addEvent(Integer id, String gameName, Double gamePrice, Integer freeSeats) throws BasketException {
        Event newEvent = new Event(id, gameName, gamePrice, freeSeats);
        this.eventRepository.save(newEvent);
        notifyObservers();
    }

    @Override
    public void deleteEvent(Integer id) throws BasketException {
        this.eventRepository.delete(id);
        notifyObservers();
    }

    @Override
    public void updateEvent(Integer id, String gameName, Double gamePrice, Integer freeSeats) throws BasketException {
        Event newEvent = new Event(id, gameName, gamePrice, freeSeats);
        this.eventRepository.update(id, newEvent);
        notifyObservers();
    }

    @Override
    public void addTicket(Integer idGame, Integer reservedSeats, String clientName) throws BasketException {
        Event newEvent = eventRepository.findOne(idGame);
        Ticket ticket = new Ticket(idGame, reservedSeats, clientName, newEvent.getGamePrice() * reservedSeats);

        if (newEvent.getFreeSeats() - ticket.getReservedSeats() >= 0) {
            newEvent.setFreeSeats(newEvent.getFreeSeats() - ticket.getReservedSeats());
            updateEvent(newEvent.getId(), newEvent.getGameName(), newEvent.getGamePrice(), newEvent.getFreeSeats());

            ticketRepository.save(ticket);
        }
    }

    @Override
    public void deleteTicket(Integer id) throws BasketException {
        Ticket ticket = ticketRepository.findOne(id);
        Event newEvent = eventRepository.findOne(ticket.getId_game());
        newEvent.setFreeSeats(newEvent.getFreeSeats() + ticket.getReservedSeats());
        ticketRepository.delete(id);
        updateEvent(newEvent.getId(), newEvent.getGameName(), newEvent.getGamePrice(), newEvent.getFreeSeats());
    }

    @Override
    public Ticket getTicket(Integer id) throws BasketException {
        return ticketRepository.findOne(id);
    }

    @Override
    public Event getEvent(Integer id) throws BasketException {
        return eventRepository.findOne(id);
    }

    private synchronized void notifyObservers() {
        for (IObserver Client : loggedPers.values()) {
            if (Client != null) {
                System.out.println("Notifying [" + Client.toString());
                try {
                    Client.update((List<Event>) getEvents());
                } catch (BasketException | RemoteException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}
